[Merged by Bors] - feat(topology/ωCPO): define Scott topology in connection with ω-complete partial orders

cipher1024 opened the pull request:
---
<!-- put comments you want to keep out of the PR commit here -->

cipher1024 committed:
feat(topology/Scott): Definition of Scott topology

cipher1024 committed:
feat(pfun/recursion): unbounded recursion (#3778)

Co-authored-by: Mario Carneiro <di.gama@gmail.com>

cipher1024 committed:
Merge branch 'master' into unbounded-recursion-scott-topology

awainverse committed:
refactor(field_theory/minimal_polynomial, *): make `aeval`, `is_integral`, and `minimal_polynomial` noncommutative (#4001)

Makes `aeval`, `is_integral`, and `minimal_polynomial` compatible with noncommutative algebras
Renames `eval₂_ring_hom_noncomm` to `eval₂_ring_hom'`

robertylewis commented on src/data/nat/up.lean:
```diff
@@ -0,0 +1,58 @@
+/-
```
Didn't you already add this file?

digama0 replied:
It was renamed to `nat.upto`. Apparently that didn't get merged back into Simon's other work.

digama0 commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,212 @@
+/-- The characteristic function of open sets is monotone and preserves
+the limits of chains. -/
+def is_open (s : set α) : Prop :=
+∃ (h : monotone s), ∀ (c : chain α) x, is_ωSup c x → is_ωSup (c.map ⟨s,h⟩) (s x)
```
This is stated in a very weird way. It took me a while to figure out what `monotone s` is supposed to mean given that `s` is a set, not a function. I guess you are abusing the defeq here to mean "`s` is upward closed", but I would prefer if you said that directly. Also there is an occurence of `s x` in this definition which is also an abuse of set defeq.

arthuraa replied:
I agree that using `set` is confusing, but I like the spirit of this formulation. I think it would be even clearer to define an open set as being the inverse image of `True` by a function `\alpha -> Prop` that is continuous in the order-theoretic sense.

cipher1024 replied:
What would be the relation between the function and the set? Also, how do you formulate the order theoretic continuity?

arthuraa replied:
I gather from the way this definition is stated that `set a` and `a -> Prop` are the same thing. Thus, the function I was talking about and the set would be the same thing. By "order-theoretic continuity", I meant that this function should satisfy the `omega_complete_partial_order.continuous'` predicate of the other file. (Note that `Prop` is a CPO for the order you are using here.)

cipher1024 replied:
If we use `continuous'` though, we need `a` to be a CPO as well. Wouldn't that be sort of begging the question?

sgouezel committed:
refactor(geometry/manifold): use a sigma type for the total space of the tangent bundle (#3966)

Redefine the total space of the tangent bundle to be a sigma type instead of a product type. Before
```
have p : tangent_bundle I M := sorry,
rcases p with ⟨x, v⟩,
-- x: M
-- v: E
```
After
```
have p : tangent_bundle I M := sorry,
rcases p with ⟨x, v⟩,
-- x: M
-- v: tangent_space I x
```
This seems more natural, and is probably needed to do Riemannian manifolds right. The drawback is that we can not abuse identifications any more between the tangent bundle to a vector space and a product space (but we can still identify the tangent space with the vector space itself, which is the most important thing). 


Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

arthuraa commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,212 @@
+    apply ωSup_le, }
+end
+
+lemma continuous_of_Scott_continuous {α β}
```
If you defined an open set literally as the preimage of `True` by an order-continuous function `X -> Prop`, then this theorem should just be an application of the fact that order-continuous functions are closed under composition.

cipher1024 committed:
delete old file

cipher1024 committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into unbounded-recursion-scott-topology

cipher1024 committed:
fix unfolding in `mono`

cipher1024 committed:
redefine `is_open` to get it closer to `continuous'`

cipher1024 committed:
fix build

cipher1024 committed:
fix topology

cipher1024 committed:
address `lint` issues

cipher1024 committed:
Merge branch 'master' into unbounded-recursion-scott-topology

jcommelin commented on src/order/basic.lean:
```diff
@@ -440,3 +440,14 @@ variables {s : β → β → Prop} {t : γ → γ → Prop}
 noncomputable def classical.DLO (α) [LO : linear_order α] : decidable_linear_order α :=
 { decidable_le := classical.dec_rel _, ..LO }
+
+/-- Type gadget create an instance of `linear_order` from a
```
```suggestion
/-- Type synonym to create an instance of `linear_order` from a
```

jcommelin commented on src/order/bounded_lattice.lean:
```diff
@@ -290,6 +290,10 @@ instance bounded_distrib_lattice_Prop : bounded_distrib_lattice Prop :=
   bot          := false,
   bot_le       := @false.elim }
 
+instance linear_order_Prop : linear_order Prop :=
```
In which namespace are we? I would expect this instance to have name `_root_.Prop.linear_order`.

cipher1024 replied:
We're in root. I'll rename the definition.

jcommelin commented on src/order/bounded_lattice.lean:
```diff
@@ -338,33 +342,41 @@ instance pi.has_top {ι : Type*} {α : ι → Type*} [Π i, has_top (α i)] : ha
   (⊤ : Π i, α i) i = ⊤ :=
 rfl
 
+@[simps]
```
Lovely use of `simps` is lovely

jcommelin commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,155 @@
+# Scott Topological Spaces
+
+I don't know what they're good for but their notion
+of continuity is equivalent to continuity in ωCPOs.
```
```suggestion
A type of topological spaces whose notion
of continuity is equivalent to continuity in ωCPOs.
```

cipher1024 replied:
I forgot that was the module comment I wrote :P I was assuming that would be the first thing reviewers would harp on.

jcommelin commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,155 @@
+@[reducible]
+def Scott (α : Type u) := α
+
+instance scott_topological_space (α : Type u) [omega_complete_partial_order α] : topological_space (Scott α) :=
```
```suggestion
instance Scott.topological_space (α : Type u) [omega_complete_partial_order α] : topological_space (Scott α) :=
```

jcommelin commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,155 @@
+open_locale classical
+
+universes u
+namespace scott_topological_space
```
```suggestion
namespace Scott
```

jcommelin commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,155 @@
+  apply complete_lattice.Sup_continuous' _ h₀,
+end
+
+end scott_topological_space
```
```suggestion
end Scott
```

jcommelin commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,155 @@
+  is_open_sUnion := scott_topological_space.is_open_sUnion α }
+
+@[simp]
+lemma set_of_apply {α} (p : α → Prop) (x : α) : set_of p x ↔ p x := iff.rfl
```
Should this be in this file?

digama0 replied:
This should not exist at all, it is evidence of abstraction breaking.

jcommelin commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,155 @@
+lemma set_of_apply {α} (p : α → Prop) (x : α) : set_of p x ↔ p x := iff.rfl
+
+@[simp]
+lemma le_iff_imp {p q : Prop} : p ≤ q ↔ (p → q) := by refl
```
Idem

jcommelin reviewed (commented):
Some minor comments

digama0 commented on src/topology/omega_complete_partial_order.lean:
```diff
@@ -0,0 +1,155 @@
+/-- The characteristic function of open sets is monotone and preserves
+the limits of chains. -/
+def is_open (s : set α) : Prop :=
+continuous' (s : α → Prop)
```
The coercion here is still breaking the abstraction. Instead use `continuous' (\lam a, a \in s)` . (Not sure if this comes up elsewhere, but the reverse coercion should be written `{x | p x}`.)

cipher1024 replied:
Would it be work changing the `set` definition in core to make it `irreducible`?

cipher1024 committed:
make `set` locally irreducible and convert explicitly between a set
and its characteristic function

cipher1024 committed:
Apply suggestions from code review

Co-authored-by: Johan Commelin <johan@commelin.net>

cipher1024 committed:
remove unrelated lemmas

cipher1024 committed:
little fixes

jcommelin reviewed (approved):
Thanks :tada:

bors merge
